algorithms/src/polycommit/sonic_pc/data_structures.rs

```diff
@@ -387,6 +387,68 @@ pub struct VerifierKey<E: PairingEngine> {
     pub max_degree: usize,
 }
 
+impl<E: PairingEngine> CanonicalSerialize for VerifierKey<E> {
+    fn serialize_with_mode<W: Write>(&self, mut writer: W, compress: Compress) -> Result<(), SerializationError> {
+        self.vk.serialize_with_mode(&mut writer, compress)?;
+        self.degree_bounds_and_neg_powers_of_h.serialize_with_mode(&mut writer, compress)?;
+        self.supported_degree.serialize_with_mode(&mut writer, compress)?;
+        self.max_degree.serialize_with_mode(&mut writer, compress)?;
+        Ok(())
+    }
+
+    fn serialized_size(&self, compress: Compress) -> usize {
+        self.vk.serialized_size(compress)
+            + self.degree_bounds_and_neg_powers_of_h.serialized_size(compress)
+            + self.supported_degree.serialized_size(compress)
+            + self.max_degree.serialized_size(compress)
+    }
+}
+
+impl<E: PairingEngine> CanonicalDeserialize for VerifierKey<E> {
+    fn deserialize_with_mode<R: Read>(
+        mut reader: R,
+        compress: Compress,
+        validate: Validate,
+    ) -> Result<Self, SerializationError> {
+        let vk = CanonicalDeserialize::deserialize_with_mode(&mut reader, compress, validate)?;
+        let degree_bounds_and_neg_powers_of_h: Option<Vec<(usize, E::G2Affine)>> =
+            CanonicalDeserialize::deserialize_with_mode(&mut reader, compress, validate)?;
+        let supported_degree = CanonicalDeserialize::deserialize_with_mode(&mut reader, compress, validate)?;
+        let max_degree = CanonicalDeserialize::deserialize_with_mode(&mut reader, compress, validate)?;
+        let degree_bounds_and_prepared_neg_powers_of_h =
+            degree_bounds_and_neg_powers_of_h.as_ref().map(|v| v.iter().map(|(b, pow)| (*b, pow.prepare())).collect());
```

Comment:
This is saves a lot of space, but is worth benching the deserializations to explicitly measure the cost of recomputing the G2 preparation? If it ends up being particularly heavy, what effect might this have on verifying nodes or application implementors?

Reply:
Good idea. I've added benchmarks in the latest commit, and updated the PR description to report some preliminary numbers